### PR TITLE
Add optimized percolator chapter by @MyonKeminta

### PR DESCRIPTION
**Pull request by Hoverbear**
This is a PR with content from @MyonKeminta that was posted on https://github.com/tikv/deep-dive-tikv/pull/32.

### CODE REVIEW

File: content/deep-dive/distributed-transaction/optimized-percolator.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+commit_ts = max{start_ts, region_1_max_read_ts, region_2_max_read_ts, ...} + 1
+```
+
+where `region_N_max_read_ts` is the maximum timestamp of all reads on the
+region, for all regions involved in the transaction.
+
+## Single Region 1PC
+
+{{< warning >}}
+This optimization haven't been finished yet, but will be available in the future.
```
Comment (Hoverbear) [inline content/deep-dive/distributed-transaction/optimized-percolator.md]: Do we have any tracking issues for this @MyonKeminta ?

File: content/deep-dive/distributed-transaction/optimized-percolator.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+## Point Get Without Timestamp
+
+Timestamps are critical to provide the isolation. For every transaction, we allocate a unique `start_ts` for it, and ensures a transaction T can only see data committed before T's `start_ts`.
+
+But if a transaction T does nothing but reads a single key, is it really necessary to allocate it a `start_ts`? The answer is no. We can simply read newest version directly, because it's equivalent to reading with `start_ts` which is exactly the instant when the key is read. It's even ok to read a locked key, because it's equivalent to reading with `start_ts` allocated before the lock's `start_ts`.
+
+## Calculated Commit Ts
+
+{{< warning >}}
+This optimization haven't been finished yet, but will be available in the future. [RFC](https://github.com/tikv/rfcs/pull/25).
```
Comment (Hoverbear) [inline content/deep-dive/distributed-transaction/optimized-percolator.md]: Do we have any tracking issues for this @MyonKeminta ?

**Comment by MyonKeminta** [comment]
Not yet. I'm sorry 😢

**Comment by Hoverbear** [comment]
OK  :)

**Comment by Hoverbear** [comment]
@MyonKeminta PTAL

File: content/deep-dive/distributed-transaction/optimized-percolator.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ahead of the actual time if we don't ask PD.
+
+To conclude, in order not to break the Snapshot Isolation and the integrity, a
+valid range for `commit_ts` should be:
+
+```text
+max{start_ts, max_read_ts_of_written_keys} < commit_ts <= now
+```
+
+So here comes an idea to calculate the commit_ts:
```
Comment (dcalvin) [inline content/deep-dive/distributed-transaction/optimized-percolator.md]: "idea" is too general and causal  
```suggestion
So here comes a method to calculate the commit_ts:
```

File: content/deep-dive/distributed-transaction/optimized-percolator.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+where `region_N_max_read_ts` is the maximum timestamp of all reads on the
+region, for all regions involved in the transaction.
+
+## Single Region 1PC
+
+{{< warning >}}
+This optimization haven't been finished yet, but will be available in the future.
+{{</ warning >}}
+
+For non-distributed databases, it's easy to provide ACID transactions; but for distributed databases, usually 2PC(two-phase commit) is required to make transactions ACID. Percolator provides such a 2PC algorithm, which is adopted by TiKV.
```
Comment (dcalvin) [inline content/deep-dive/distributed-transaction/optimized-percolator.md]:
```suggestion
For non-distributed databases, it's easy to provide ACID transactions; but for distributed databases, usually 2PC (two-phase commit) is required to make transactions ACID. Percolator provides such a 2PC algorithm, which is adopted by TiKV.
```

File: content/deep-dive/distributed-transaction/optimized-percolator.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+There's an optimization for short values to avoid touching `CF_DEFAULT` in TiKV. If the value is short enough, the value will not be put into `CF_DEFAULT` during prewrite phase. Instead, it will be inlined in the lock and saved in `CF_LOCK`. Then in commit phase, the value will be moved out from the lock and inlined in the write record. Therefore, we do not need to touch `CF_DEFAULT` to access and manipulate short values.
+
+## Point Get Without Timestamp
+
+Timestamps are critical to provide the isolation. For every transaction, we allocate a unique `start_ts` for it, and ensures a transaction T can only see data committed before T's `start_ts`.
+
+But if a transaction T does nothing but reads a single key, is it really necessary to allocate it a `start_ts`? The answer is no. We can simply read newest version directly, because it's equivalent to reading with `start_ts` which is exactly the instant when the key is read. It's even ok to read a locked key, because it's equivalent to reading with `start_ts` allocated before the lock's `start_ts`.
+
+## Calculated Commit Ts
```
Comment (dcalvin) [inline content/deep-dive/distributed-transaction/optimized-percolator.md]:
```suggestion
## Calculated Commit Timestamp
```

File: content/deep-dive/distributed-transaction/optimized-percolator.md (hunk 2 above)
Comment (dcalvin) [inline content/deep-dive/distributed-transaction/optimized-percolator.md]:
```suggestion
This optimization hasn't been finished yet, but will be available in the future. [RFC](https://github.com/tikv/rfcs/pull/25).
```

File: content/deep-dive/distributed-transaction/optimized-percolator.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+## Short Value in Write Column
+
+As said before, TiKV uses RocksDB's column families to save different columns of Percolator. Different column families of RocksDB is actually different LSM-Trees. When we accessing a value we need to first search the `CF_WRITE` to find next record's `start_ts`, and then search the corresponding record in `CF_DEFAULT`. If a value is very small, it is a waste to search RocksDB twice.
+
+There's an optimization for short values to avoid touching `CF_DEFAULT` in TiKV. If the value is short enough, the value will not be put into `CF_DEFAULT` during prewrite phase. Instead, it will be inlined in the lock and saved in `CF_LOCK`. Then in commit phase, the value will be moved out from the lock and inlined in the write record. Therefore, we do not need to touch `CF_DEFAULT` to access and manipulate short values.
+
+## Point Get Without Timestamp
```
Comment (dcalvin) [inline content/deep-dive/distributed-transaction/optimized-percolator.md]:
```suggestion
## Point Read Without Timestamp
```

**Comment by dcalvin** [comment]
@Hoverbear Please could you make the suggested changes? I am good with the rest.

**Comment by Hoverbear** [comment]
Sorry @dcalvin , this fell by the wayside.

File: content/deep-dive/distributed-transaction/optimized-percolator.md (hunk 1 above)
Comment (Hoverbear) [reply content/deep-dive/distributed-transaction/optimized-percolator.md]: Ping @MyonKeminta

File: content/deep-dive/distributed-transaction/optimized-percolator.md (hunk 2 above)
Comment (Hoverbear) [reply content/deep-dive/distributed-transaction/optimized-percolator.md]: Ping @MyonKeminta

**Comment by MyonKeminta** [comment]
@Hoverbear Sorry the two new optimizations are not tracked in GitHub issues now... and for some reason they are blocked. So I'm not sure whether these two optimizations can be stated here.

**Comment by MyonKeminta** [comment]
https://github.com/tikv/rfcs/pull/25 Here is an RFC for calculated commit ts but it hasn't been merged yet

**Pull request merged**